rp2: Force available data to the ring buffer on read.

robert-hh opened the pull request:
So uart.read() will return single characters even with a setting of timeout=0.

dpgeorge commented on ports/mimxrt/mimxrt_flash.c:
```diff
@@ -82,16 +82,27 @@ status_t flash_erase_block(uint32_t erase_addr) {
 status_t flash_write_block(uint32_t dest_addr, const uint8_t *src, uint32_t length) __attribute__((section(".ram_functions")));
 status_t flash_write_block(uint32_t dest_addr, const uint8_t *src, uint32_t length) {
     status_t status;
+    uint32_t size;
```
I think the change to this function is not supposed to be in this PR

robert-hh replied:
Surely not. That happens when working at two branches at the same time.

dpgeorge commented on ports/rp2/machine_uart.c:
```diff
@@ -404,6 +404,11 @@ STATIC mp_uint_t machine_uart_read(mp_obj_t self_in, void *buf_in, mp_uint_t siz
+    // Force a few incoming bytes to the buffer
+    self->read_lock = true;
+    uart_drain_rx_fifo(self);
+    self->read_lock = false;
```
Might this lead to lost characters, if the ringbuf is already full and there are bytes in the FIFO?

robert-hh replied:
I'll change that.

robert-hh committed:
rp2: Force available data to the ring buffer on read.

But only if the ring buffer is empty.

dpgeorge commented:
@iabdalkader are you able to test this in conjunction with your ioctl change, and your BLE work?

iabdalkader commented:
Yes I can test it with ioctl fix and BLE, but wouldn't this also fix the issue without repeating code ?

```diff
diff --git a/ports/rp2/machine_uart.c b/ports/rp2/machine_uart.c
index 2431f496e..dc2a9edbd 100644
--- a/ports/rp2/machine_uart.c
+++ b/ports/rp2/machine_uart.c
@@ -405,6 +405,7 @@ STATIC mp_uint_t machine_uart_read(mp_obj_t self_in, void *buf_in, mp_uint_t siz
     uint8_t *dest = buf_in;
 
     for (size_t i = 0; i < size; i++) {
+        t = time_us_64() + timeout_char_us;
         // Wait for the first/next character
         while (ringbuf_avail(&self->read_buffer) == 0) {
             if (time_us_64() > t) {  // timed out
@@ -422,7 +423,6 @@ STATIC mp_uint_t machine_uart_read(mp_obj_t self_in, void *buf_in, mp_uint_t siz
             self->read_lock = false;
         }
         *dest++ = ringbuf_get(&(self->read_buffer));
-        t = time_us_64() + timeout_char_us;
     }
     return size;
 }
```

It will allow the inner loop to try to drain the RX FIFO at least once before timing out.

robert-hh commented:
That would work, since timeout_char must not be 0. But the first t value is set at the start of the function and is the timeout for the first character. timeout_char is the timeout for the following characters. So this statement must not be moved.

iabdalkader commented:
> is the timeout for the first character

That's right, okay how about this change ? It should have the same effect right ?

```diff
diff --git a/ports/rp2/machine_uart.c b/ports/rp2/machine_uart.c
index 2431f496e..a8ec149f4 100644
--- a/ports/rp2/machine_uart.c
+++ b/ports/rp2/machine_uart.c
@@ -407,6 +407,13 @@ STATIC mp_uint_t machine_uart_read(mp_obj_t self_in, void *buf_in, mp_uint_t siz
     for (size_t i = 0; i < size; i++) {
         // Wait for the first/next character
         while (ringbuf_avail(&self->read_buffer) == 0) {
+            if (uart_is_readable(self->uart)) {
+                // Force a few incoming bytes to the buffer
+                self->read_lock = true;
+                uart_drain_rx_fifo(self);
+                self->read_lock = false;
+                break;
+            }
             if (time_us_64() > t) {  // timed out
                 if (i <= 0) {
                     *errcode = MP_EAGAIN;
@@ -416,10 +423,6 @@ STATIC mp_uint_t machine_uart_read(mp_obj_t self_in, void *buf_in, mp_uint_t siz
                 }
             }
             MICROPY_EVENT_POLL_HOOK
-            // Force a few incoming bytes to the buffer
-            self->read_lock = true;
-            uart_drain_rx_fifo(self);
-            self->read_lock = false;
         }
         *dest++ = ringbuf_get(&(self->read_buffer));
         t = time_us_64() + timeout_char_us;
```

@dpgeorge I tested BLE with ioctl fix and both fixes (this PR and the suggestion abouve) with timeout==0 and both are working fine. At least BLE is working fine, that much I can confirm.

robert-hh commented:
Not exactly. Because if the ringbuffer is already full when entering the function, calling the drain function would case these to get lost. And it would not fix the  timeout=0 issue.

iabdalkader commented:
> Because if the ringbuffer is already full when entering the function, calling the drain function would case these to get lost

If it's full when entering it won't drain the rx fifo, t because it's checked first:
```
while (ringbuf_avail(&self->read_buffer) == 0) {
```

robert-hh commented:
Yes. But it would still leave early for timeout=0.

iabdalkader commented:
> Yes. But it would still leave early for timeout=0.

If the ringbuf is empty *and* the RX FIFO is empty too, it will return EAGAIN immediately with timeout == 0 which is the right behavior. The RX FIFO is checked and drained at least once before returning EAGAIN with no extra delay.


>  Edit: The test for uart being readable is not required, because that happens in the drain function anyhow.

It's required so that it breaks if there were any characters in the FIFO (meaning the ringbuf now has 1 ) otherwise it would return EAGAIN even if it did drain the fifo.

robert-hh commented:
Right. I did not notice the break statement. Using uart_is_readable() is slightly faster then.

robert-hh committed:
rp2: Resortering machine_uart_read() code.

Following the suggestion of @iabdalkader. Slightly shorter and somewhat
slower.

dpgeorge commented:
Thanks for working this out and testing.  Merged in ee49ae8f82c22b43777aa9e73a1b7a6825304d13
